crypto/fipsmodule/ec/ec_nistp.c

```diff
@@ -498,3 +498,141 @@ void ec_nistp_scalar_mul(const ec_nistp_meth *ctx,
+void ec_nistp_scalar_mul_base(const ec_nistp_meth *ctx,
+                              ec_nistp_felem_limb *x_out,
+                              ec_nistp_felem_limb *y_out,
+                              ec_nistp_felem_limb *z_out,
+                              const EC_SCALAR *scalar) {
+  // Regular-wNAF encoding of the scalar.
+  int16_t rwnaf[SCALAR_MUL_MAX_NUM_WINDOWS];
+  scalar_rwnaf(rwnaf, SCALAR_MUL_WINDOW_SIZE, scalar, ctx->felem_num_bits);
+  size_t num_windows = DIV_AND_CEIL(ctx->felem_num_bits, SCALAR_MUL_WINDOW_SIZE);
+
+  // We need two point accumulators, so we define them of maximum size
+  // to avoid allocation, and just take pointers to individual coordinates.
+  // (This cruft will disapear when we refactor point_add/dbl to work with
+  // whole points instead of individual coordinates).
+  ec_nistp_felem_limb res[3 * FELEM_MAX_NUM_OF_LIMBS] = {0};
+  ec_nistp_felem_limb tmp[3 * FELEM_MAX_NUM_OF_LIMBS] = {0};
+  ec_nistp_felem_limb *x_res = &res[0];
+  ec_nistp_felem_limb *y_res = &res[ctx->felem_num_limbs];
+  ec_nistp_felem_limb *z_res = &res[ctx->felem_num_limbs * 2];
+  ec_nistp_felem_limb *x_tmp = &tmp[0];
+  ec_nistp_felem_limb *y_tmp = &tmp[ctx->felem_num_limbs];
+  ec_nistp_felem_limb *z_tmp = &tmp[ctx->felem_num_limbs * 2];
+
+  // Process the 4 groups of digits starting from group (3) down to group (0).
+  for (int i = 3; i >= 0; i--) {
+    // Double |res| 5 times in each iteration, except in the first one.
+    for (size_t j = 0; i != 3 && j < SCALAR_MUL_WINDOW_SIZE; j++) {
+      ctx->point_dbl(x_res, y_res, z_res, x_res, y_res, z_res);
+    }
+
+    // Process the digits in the current group from the most to the least
+    // significant one.
+    size_t start_idx = ((num_windows - i - 1) / 4) * 4 + i;
+
+    for (int j = start_idx; j >= 0; j -= 4) {
+      // For each digit |d| in the current group read the corresponding point
+      // from the table and add it to |res|. If |d| is negative, negate
+      // the point before adding it to |res|.
+      int16_t d = rwnaf[j];
+      int16_t is_neg = (d >> 15) & 1; // is_neg = (d < 0) ? 1 : 0
+      d = (d ^ -is_neg) + is_neg;     // d = abs(d)
+
+      int16_t idx = d >> 1;
+
+      // Select the point to add, in constant time.
+      size_t point_num_limbs = 2 * ctx->felem_num_limbs;  // Affine points.
+      size_t subtable_num_limbs = SCALAR_MUL_TABLE_NUM_POINTS * point_num_limbs;
+      size_t table_idx = (j / 4) * subtable_num_limbs;
+      const ec_nistp_felem_limb *table = &ctx->scalar_mul_base_table[table_idx];
+      select_point_from_table(ctx, tmp, table, idx, 0);
+
+      // Negate y coordinate of the point tmp = (x, y); ftmp = -y.
+      ec_nistp_felem ftmp;
+      ctx->felem_neg(ftmp, y_tmp);
+
+      cmovznz(y_tmp, ctx->felem_num_limbs, is_neg, y_tmp, ftmp);
+
+      // Add the point to the accumulator |res|.
+      ctx->point_add(x_res, y_res, z_res, x_res, y_res, z_res, 1,
+                     x_tmp, y_tmp, ctx->felem_one);
```

Comment:
Should we retain the documentation around `p384/521_point_add`?

Reply:
not needed imo, but if you insist I'll add it (maybe in the next pr if no other changes are needed for this one).